File: apps/api/app/models/annotation.py
"""Annotation pair models for label-bbox pairing persistence."""

from datetime import datetime

from pydantic import BaseModel


class AnnotationBBox(BaseModel):
    """Normalized 0-1 bounding box coordinates."""

    x: float
    y: float
    width: float
    height: float


class AnnotationPairModel(BaseModel):
    """Full annotation pair as stored in the database."""

    id: str
    document_id: str
    label_id: str
    label_text: str
    label_bbox: AnnotationBBox
    label_page: int
    field_id: str
    field_name: str
    field_bbox: AnnotationBBox
    field_page: int
    confidence: float = 100.0
    status: str = "confirmed"
    is_manual: bool = True
    created_at: datetime | None = None


class AnnotationPairCreate(BaseModel):
    """Request body for creating an annotation pair."""

    label_id: str
    label_text: str
    label_bbox: AnnotationBBox
    label_page: int
    field_id: str
    field_name: str
    field_bbox: AnnotationBBox
    field_page: int
    confidence: float = 100.0
    status: str = "confirmed"
    is_manual: bool = True


class AnnotationPairsResponse(BaseModel):
    """Response wrapper for a list of annotation pairs."""

    pairs: list[AnnotationPairModel]

File: apps/web/src/types/annotation.ts
/**
 * Type definitions for the Label & Bbox Annotation Tool.
 * Used to pair text labels with form field bounding boxes.
 */

export interface AnnotationBBox {
  /** Normalized 0-1 coordinate */
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface LabelOverlay {
  id: string;
  text: string;
  bbox: AnnotationBBox;
  page: number;
}

export interface FieldOverlay {
  id: string;
  fieldName: string;
  bbox: AnnotationBBox;
  page: number;
}

export interface AnnotationPair {
  id: string;
  label: LabelOverlay;
  field: FieldOverlay;
  confidence: number;
  status: 'confirmed' | 'flagged';
  isManual: boolean;
}

export type AnnotationMode =
  | { type: 'idle' }
  | { type: 'field-selection'; labelId: string }
  | { type: 'focus-pair'; labelId: string; fieldId: string; pairId: string };

export type LabelColorKey = 'default' | 'paired' | 'focused';
export type FieldColorKey = 'default' | 'paired';

export interface OverlayConfig {
  labels: Array<{ overlay: LabelOverlay; colorKey: LabelColorKey }>;
  fields: Array<{ overlay: FieldOverlay; colorKey: FieldColorKey }>;
}

export interface AnnotationDocumentState {
  documentId: string;
  filename: string;
  totalPages: number;
  currentPage: number;
  labels: LabelOverlay[];
  fields: FieldOverlay[];
  pairs: AnnotationPair[];
}

export interface ExportEntry {
  label_text: string;
  label_bbox: AnnotationBBox;
  field_bbox: AnnotationBBox;
  field_id: string;
  field_name: string;
  page: number;
  confidence: number;
  status: 'confirmed' | 'flagged';
}
